article/results.tex

```diff
@@ -56,59 +56,75 @@ \subsection{Best Practice Guidelines}
 As part of this work we have contributed substantial changes to the original OPFVTA repository, based on which we formulate a number of best practice guidelines, highly relevant in the production of reexecutable research outputs.
 
 \subsubsection{Errors should be fatal more often than not}
-The high complexity of a full analysis workflow makes missing output data intractable, as there are many steps at which data may have been dropped.
-Manually setting individual invocations propagate exit codes throughout the Make system is cumbersome, and thus an easy global invocation should be relied on so that errors become fatal and immediately visible.
-This is best accomplished by prepending all Bash scripts for which this concern is relevant with a \texttt{set -eu} line.
+
+By default any Python script or a compiled C or C++ program just errors out and exits immediately when running into some unexpected operation.
+Shell scripts behave differently.
+By default they continue with execution of the next scripted command, and only when the list of commands is exhausted or explicit `exit` command is called, it finally exits with a non-0 error code.
+As a result, even if not causing some detrimental actions, an execution of the script could continue for hours before it fails, and original error message might be lost in the flood of output making it hard or impossible to localize original problem to be fixed.
+Analogously, shell scripts treat undefined variables as a variable having an empty value, and thus not causing any error at all.
```

Comment:
Is this true? I'm pretty sure bash distinguishes empty from unset, is undefined different from unset?

Reply:
try and show me wrong (which I can well  be)

Reply:
```
[deco]~ ❱ if [ -v FIFI ]; then echo "a" ; fi
[deco]~ ❱ FIFI=""; if [ -v FIFI ]; then echo "a" ; fi
a
```

Reply:
note - `-v` is bashism:

```shell
❯ sh -c '[ -v FIFI ] && echo a'
sh: 1: [: -v: unexpected operator
```

and what you show above is that `-v` behaves as it is documented:

```
       -v varname
              True if the shell variable varname is set (has been assigned a value).
```

and not "bash distinguishes empty from unset," which would be `-z` or `-n` depending on what bool you want out of it.  So "bash distinguishes empty from unset" is an undefined statement by itself -- bash does have operands allowing to figure out all those cases.

But what matters here is not what bash can do, but how it behaves! When variable is not defined, shell doesn't care and just uses empty value for it:

```shell
❯ sh -c 'echo ">$FIFI<"; echo a'
><
a
❯ sh -c 'FIFI=;echo ">$FIFI<"; echo a'
><
a
```

**unless** you `set -u`,

```shell
❯ sh -c 'set -u; FIFI=;echo ">$FIFI<"; echo a'
><
a
❯ sh -c 'set -u; echo ">$FIFI<"; echo a'
sh: 1: FIFI: parameter not set
```

Reply:
ok.